### ISSUE TITLE
CGE doesn't compile with Delphi XE4

### ISSUE THREAD

**Issue by peardox** [castle-engine/castle-engine#481]
Don't know if you wanna go back this far either so about to try XE7 (as I know it has some stuff missing below)

1) no function StrToUInt64
2) no type PGLenum (from 4)
3) no const CP_UTF8 (from 5)
4) No unit OpenGLexl
5) No unit System.NetEncoding - more issues here than reported

There may be more but I'm saving workstate and switching to XE7 until I hear if you want more exploration for XE4

**Comment by michaliskambi** [comment]
Thank you! 

I do not have a decision yet whether we'll go with XE4 or XE7, but the list of XE4 lacks seems workable.

-  System.NetEncoding missing - we use it only for one, but important, place: `TBase64DecodingStream`. Though I guess if XE4 doesn't have built-in base64 decoding routines, we could even roll our own, this is not a big thing.

- OpenGLext missing - I have been already thinking of using different OpenGL headers, not OpenGL/OpenGLExt. So it would solve this issue too. See https://github.com/castle-engine/castle-engine/issues/334 -- either dglOpenGL (already available) or GLAD or GLAD 2. I'm most enthusiastic about using GLAD 2 (because it will work for all future versions, for Vulkan too) but it needs work, right now it doesn't support Pascal. I'll make a "call for help" about it in the news ("do we have volunteers interested in extending GLAD 2 to support Pascal") and see what happens.

**Comment by peardox** [comment]
StrToUInt64 was fixable by copying from D11.3 and making a very minor mod

I stopped the compiler complaining about OpenGLext simply by copying the D11.3 code into a temp 'see if it works' and it was fine after also adding PGLEnum = ^GLEnum; to the top of the temp fix (there are actually more - copied them all and it worked)

System.NetEncoding - started to temp patch - needed const CP_UTF8 but after that fix more stuff so decided to try XE7 (about ready to start on that one now)

Basically I'm trying to identify which compilers CURRENTLY work and XE7 seems likely

**Comment by michaliskambi** [comment]
I marked this and 3 other issues about older Delphis with "backlog - undecided" because it is literally not decided yet whether we'll really push to make CGE compatible with these older Delphi versions. 

We are comfortable for now claiming support only for Delphi >= 10.2.1 , per https://castle-engine.io/delphi#_delphi_versions_supported .

Reasons why I may just decide to resign from supporting older Delphi versions:

- While it seems easy to do to add some small fixes for older Delphis, it would also be a constant burden to recheck all we do with earlier Delphi versions. 

- Aside from one issue on GitHub ( https://github.com/castle-engine/castle-engine/issues/475 ) , we didn't really have instances of people asking for support for older Delphi versions. Well, aside from occasional questions to support Delphi 7 :), which isn't really possible. So the need to support e.g. Delphi XE4, based on actual user feedback, is small (well, actually zero, as far as is measured; but I of course recognize that most users would just "go away without asking" when they see their old Delphi version is not supported, and only a minority would ask us about it.)

- Cross-platform support, i.e. anything aside from Windows, is anyway not really doable with these older Delphi versions.  

    Delphi < 10.4 for mobile have weird String treatment making it impossible to support.  Details in https://castle-engine.io/delphi#_delphi_versions_supported

    Delphi < 11.3 has now broken GetIt (so getting FMXLinux is hard), for over a month (since January 17th -- https://blogs.embarcadero.com/we-are-experiencing-a-hardware-outage/ ) and I'm unsure if it will be ever really restored. It seems it's not a priority for Embarcadero -- they already took a long time to fix GetIt for Delphi 12, then it took longer for 11.3. There's no ETA for older versions.

    So older Delphis would likely be only supported on Windows only, which makes them even less useful for many projects.

**Comment by peardox** [comment]
> Delphi < 11.3 has now broken GetIt (so getting FMXLinux is hard), for over a month (since January 17th
my.embarcadero.com DevRel has stuff like 
"FMXLinux Package Download for RAD Studio 11.3"

**Comment by michaliskambi** [comment]
> my.embarcadero.com DevRel has stuff like
> "FMXLinux Package Download for RAD Studio 11.3"

Oh I know about it (thanks to you telling me about it a while back! :) ). I also mention it on https://castle-engine.io/delphi_linux as alternative to downloading from GetIt. Still, GetIt seems much more comfortable for normal users. 

And my.embarcadero.com doesn't seem to include FMXLinux for older Delphis than 11.3. Maybe it doesn't even exist, even in GetIt? It is not possible to test it now. Making it impossible to support Delphi/Linux + CGE with older Delphis.

**Comment by peardox** [comment]
FMX Linux came bundled starting with 10.3.2
https://en.wikipedia.org/wiki/History_of_Delphi_(software)#Delphi_10.3.2_Update_2[25]

It is also availabe back to 10.2.x (for money) https://www.fmxlinux.com/order.html

**Comment by michaliskambi** [comment]
Same as https://github.com/castle-engine/castle-engine/issues/482 :

> Thank you for testing -- for now I say to officially cut these older Delphi versions. The effort isn't really trivial, as any special hacks would need to be maintained / tested, and the usefullness for users is anyway not big -- since non-Windows platforms would not be supported with these older Delphis anyway, so the big CGE feature "cross platform" would not in practice be relevant for people with these old Delphis.
